File: 114. Flatten Binary Tree to Linked List.py
T = O(n)
S = o(h)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root or (not root.right and not root.left):
            return
            
        if root.left:
            self.flatten(root.left)
            temp = root.right
            root.right = root.left
            root.left = None
            while root.right:
                root = root.right
            root.right = temp
        self.flatten(root.right)

File: 114_Flatten_Binary_Tree_to_Linked_List.java
/**
 * 
 * approach - 
 * take left tree, flaten it completely.
 * 
 * now take right tree, flatten it as well 
 * 
 * at last move to the last node of left tree, and assign the entire right treee.
 * tc:O(n) + 2pass
 * sc : O(h) 
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {

        // base case
        if (root == null)
            return;

        TreeNode ltree = root.left;
        root.left = null;

        TreeNode rtree = root.right;

        // recusrive calls
        flatten(ltree);

        // make link, with finalized left tree
        root.right = ltree;

        // recusre to right tree and flatten it
        flatten(rtree);

        // here we have flattened right part, to assign it, travese to the right most
        // child of ltree.

        TreeNode curr = root;
        while (curr.right != null) {
            curr = curr.right;
        }

        curr.right = rtree;

        return;

    }
}